cartographer/mapping_2d/sparse_pose_graph/optimization_problem.cc

```diff
@@ -192,6 +209,27 @@ void OptimizationProblem::Solve(const std::vector<Constraint>& constraints,
             .data());
   }
 
+  // Add penalties for violating odometry
+  for (size_t trajectory_id = 0; trajectory_id != odometry_data_.size();
+       ++trajectory_id) {
+    for (size_t odometry_data_index = 1;
+         odometry_data_index < odometry_data_[trajectory_id].size();
+         ++odometry_data_index) {
+      problem.AddResidualBlock(
+          new ceres::AutoDiffCostFunction<WarpedSpaCostFunction, 3, 3, 3, 2>(
+              new WarpedSpaCostFunction(Constraint::Pose{
+                  odometry_data_[trajectory_id][odometry_data_index - 1]
+                      .pose.inverse() *
+                  odometry_data_[trajectory_id][odometry_data_index]
+                      .pose,
+                  0.3, 30.})),
+          nullptr /* loss function */,
+          C_nodes[trajectory_id][odometry_data_index - 1].data(),
```

Comment:
This seems wrong: indices of nodes and odometry data are different in general since usually odometry has a higher frequency. I think what you want to do is look at the time stamps of two consecutive nodes, then use the data in `odometry_data_` to compute the relative change expected according to odometry. Since you most likely do not have the exact timestamp in the odometry data, you'd have to interpolate between two odometry poses, once for the timestamp of the first node and once for the timestamp of the second node.